Add ALSA-based volume control and test logic

WasabiFan opened the pull request:
This PR adds an "Audio" section in the main menu and associated sub-windows to allow volume control from brickman. The primary volume page looks like this:
![volume control window](https://cloud.githubusercontent.com/assets/3310349/16213296/2f017ce0-3703-11e6-8b37-c6aa359dd5b0.png)

There is also a list window that only appears if there are no mixer elements or there is more than one mixer element. That window can be accessed from the desktop test.

That menu has the following options:
- Volume up
- Volume down
- Half volume
- Maximum volume
- Minimum volume
- Mute _(this one is disabled in certain cases, including on the EV3 -- see below.)_

This audio system is architected slightly differently from other current modules in brickman. I created a `ITestableMixerElement` interface (not a perfect name, but the best I could come up with) with two implementations: one of them is the `AlsaBackedMixerElement` and the other is the `FakeMixerElement`. The views operate on `ITestableMixerElement`s, so the controllers just instantiate whichever implementation they need and hand it to the view and the view handles the common display logic.

The views connect to the `notify` signals on elements they are given, so they update automatically whenever the underlying elements change. This is mostly used by the test implementation, but it is also how the volume percentage on the main page operates: when the user hits a volume button, the view emits a signal; the controller then updates the volume value on the base element accordingly and the view gets the property change notification and updates the label.

Some notes:
- The mixer that is configured by default on the EV3 (which may also apply to the other platforms) does not have a mute switch. This means that the mute toggle is always hidden in the volume control. While one can set the volume to zero, the speakers still emit audible clicks when sound starts playing. I have tested the mute functionality on my desktop PC and know that it works.
- The standard ALSA notation for mixer elements is the name and index separated by a comma. It turns out that the font in brickman makes commas very hard to discern after certain letters, especially _k_s (which, unfortunately, is what "playback" ends in). So, after some fiddling, I decided on putting the index in brackets before the name -- if there are other ideas for how this should be formatted I can try those out too.
- I'm pretty sure I didn't do the `CMakeLists.txt` additions correctly. Specifically, on line 139, I added `asound` after `${DEPS_LIBRARIES}` because I couldn't figure out where that variable was coming from. Guidance on where that should go would be appreciated.

WasabiFan committed:
Add ALSA-based audio volume control and test logic

WasabiFan committed:
Update debian/control for audio additions

dlech commented:
> This PR adds an "Audio" section

How about calling it "Sound" instead?

> That menu has the following options:
> - Volume up
> - Volume down
> - Half volume
> - Maximum volume
> - Minimum volume
> - Mute (this one is disabled in certain cases, including on the EV3 -- see below.)

I would rather just have volume up and down (eventually, we should just have a graphic slider, but text is OK for now). If we make the change step by 10%, then you don't really need the half/min/max. Also, if the volume is set to 0, then behind the scenes, it should activate the mute if it exists, so we don't need a separate mute option.

> I created a ITestableMixerElement interface (not a perfect name, but the best I could come up with)

It sounds like you are doing MVVM in which case I would call it IMixerElementViewModel. However, this is not how any of the other views/controllers are implemented. In other controllers, we bind properties of the data model to the view rather than binding properties of a view model in the view as you are doing here. There's nothing wrong with doing it the way you have here, but I would rather stay consistent with the existing patterns.

By using the property binding feature of GObjects, we don't have to implement a function to attach to each notify signal. We only need a translate function for special cases like converting an enum to a string.

> I added asound after ${DEPS_LIBRARIES} because I couldn't figure out where that variable was coming from. Guidance on where that should go would be appreciate

The alsa library is in pkg-config, so for now, you can add it to `pkg_check_modules` on line 120. Really some of those libraries should be separated out so that they are not used in the test version. You have to delete the cmake cache in order for it to pick up the libraries again.

WasabiFan commented:
> How about calling it "Sound" instead?

OK, I'll change that.

> I would rather just have volume up and down (eventually, we should just have a graphic slider, but text is OK for now) ... Also, if the volume is set to 0, then behind the scenes, it should activate the mute if it exists, so we don't need a separate mute option.

I think there needs to be either a mute option or a "minimum" option; having to hit the "volume down" button 10 times to mute the audio would probably get pretty annoying. How about I remove the mute button as it stands along with the half and max buttons and rename the "minimum" button to "Mute"? That way, you can still easily mute it, and it all fits on one page.

Barring a slider, I was originally hoping for something simple that would let you use the left/right buttons to modify the volume so that you didn't need to select the separate menu items. The current capabilities of ev3devKit weren't enough to lay that out without significant additions, however. We'll have to revisit this sound UI if/when someone implements a slider or other similar control.

> It sounds like you are doing MVVM...

Yes, I aimed for something closer to MVVM or MVC because it seemed like there was a fair amount of manual UI management in the controller that I wanted to avoid. Essentially, I wanted to avoid the controller having any references to `MenuItem`s and other UI elements -- the controller manages the data, and the view decides how to render those data.

~~I think I might go back and change it a bit to use the preexisting binding logic to bind the models to the views; in some cases, that might help clean up the code.~~

Edit: After looking through the code, I've decided I won't try to add bindings to replace notify signals. There isn't anywhere meaningful that it would help simplify the code. The biggest issues are with interfacing with the control panel GUI, and that is limited by Gtk+'s capabilities, not our code.

> ...in which case I would call it IMixerElementViewModel

Will-do.

> The alsa library is in pkg-config, so for now, you can add it to pkg_check_modules on line 120. 

Oh, I think I see what I did wrong -- I didn't clear the cache. After doing that, I was able to remove my other hack and it seems to be working.

WasabiFan committed:
Refine audio menu and structure

WasabiFan commented:
PR updated.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+
+namespace BrickManager {
+    public class AlsaBackedMixerElement: IMixerElementViewModel, Object {
+        private const SimpleChannelId primary_channel_id = SimpleChannelId.MONO;
```
Are you planning on changing this value? I don't see a need to. So, we don't really need a constant for it.

WasabiFan replied:
It's the equivalent of a "magic number" and is used in two places, so having it as a constant makes the code more readable and maintainable. While I don't intend to change it, I already had to change it a few times in development, so there's no reason to think I wouldn't need to change it again.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+
+        public int volume {
+            get {
+                long volume = 0;
```
don't need to initialize value when used as out parameter.

```
long volume;
```

will do.

WasabiFan replied:
Because it's an out parameter, if the code that I call doesn't set it, it will be left at whatever value it had before right? So I have it this way to ensure it's a known value if an error occurs and their code doesn't set a value.

dlech replied:
well, in this case, you should check for an error and set the value if reading `volume` after an error is valid. I also just noticed that the local variable is the same name as the property, which is not ideal.

dlech replied:
I would still like to see error checking here.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+                long min_volume, max_volume;
+                alsa_element.get_playback_volume_range(out min_volume, out max_volume);
+
+                int constrained_volume = int.min(100, int.max(0, value));
```
prefer `var` for variable declarations

WasabiFan replied:
👌

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+                alsa_element.set_playback_volume_all(rounded_volume);
+
+                bool should_mute = rounded_volume <= min_volume;
+                if(is_muted != should_mute)
```
style: space after `if` and `{}` around body.

WasabiFan replied:
Will-do

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+
+        public bool is_muted {
+            get {
+                if(!can_mute)
```
style (see previous comment)

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+                if(!can_mute)
+                    return false;
+
+                int mute_switch = 0;
```
again, no need to initialize.

WasabiFan replied:
See above; although in this case, it should actually be inverted. I'll switch that to a 1.

dlech replied:
theoretically, if there is an error, the value could be changed - the state is undefined, so better to check for error and set value if error.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+                alsa_element.set_playback_switch_all(is_muted ? 0 : 1);
+        }
+    }
+}
```
newline at end of file

WasabiFan replied:
Will-do

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+            }
+        }
+
+        protected void set_is_muted(bool is_muted) {
```
why is this not a setter of the `is_muted` property?

WasabiFan replied:
You can't add a setter to the property because the interface defines it as get-only. So, separate setters. This behavior is different from that of C# and friends.

dlech replied:
If the setters are public, why not just add `set` to the interface?

WasabiFan replied:
This method is `protected` because it's only for internal use, and other implementations of the interface won't necessarily need a setter for mute.

dlech replied:
Got it. I was confusing this with the `FakeMixerElement`

dlech commented on test/ControlPanel.vala:
```diff
@@ -147,6 +158,7 @@ namespace BrickManager {
                 file_browser_controller = new FakeFileBrowserController (builder);
                 device_browser_controller = new FakeDeviceBrowserController (builder);
                 network_controller = new FakeNetworkController (builder);
+                audio_controller = new FakeAudioController(builder);
```
style: space between function name and `(`. This applies to all new files as well.

WasabiFan replied:
👌

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+    public class AlsaBackedMixerElement: IMixerElementViewModel, Object {
+        private const SimpleChannelId primary_channel_id = SimpleChannelId.MONO;
+
+        private unowned MixerElement alsa_element;
```
no need for `private` keyword. It is implied.

WasabiFan replied:
I try to be explicit whenever there's the possibility for confusion; see [here](http://stackoverflow.com/questions/4113651/why-explicitly-write-private) and [here](http://programmers.stackexchange.com/questions/149852/what-is-the-point-of-using-the-private-access-modifier-for-c-class-members/149853). It generally makes it easier for others to read and undestand the code.

dlech replied:
There's also something to be said for being consistent. We don't use `private` anywhere else in brickman (or at least not much).

WasabiFan replied:
OK, I can change that to keep it consistent.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+
+namespace BrickManager {
+    public class AudioController : Object, IBrickManagerModule {
+        private const int VOLUME_STEP = 10;
```
no need for `private`

WasabiFan replied:
See above

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+        MixerElementSelectorWindow mixer_select_window;
+        MixerElementVolumeWindow volume_window;
+
+        public string display_name { get { return "Sound"; } }
```
It would be nice if the file name matched the display name. i.e. `SoundController`

WasabiFan replied:
Will-do

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+
+        public string display_name { get { return "Sound"; } }
+
+        protected void initialize_mixer() {
```
why is this `protected`?

WasabiFan replied:
Because it's something that any future subclasses should be able to use. At a minimum, it illustrates intent, but in the future it may actually be used.

dlech replied:
I don't see this ever being subclassed. I would consider all of the controller classes `sealed` in C# speak. But if it does ever get subclassed, you can always change it to protected if needed.

WasabiFan replied:
I don't feel strongly about it either way; I'll just default it to private and it can be dealt with later if it's ever subclassed.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+                volume_window.show_element_details = false;
+                volume_window.show();
+            }
+            else
```
style: cuddled `else`. i.e.`} else {`

WasabiFan replied:
Oh gosh, you're a monster! 😁 I'll make that change.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+
+            // Wire up handlers for volume window signals
+            volume_window.volume_up.connect(() =>
+                volume_window.current_element.volume += VOLUME_STEP);
```
reference cycle. Need to use a weak reference to `volume_window` in the body of the lambda statement.

WasabiFan replied:
Hmmm... I must not understand reference cycles well enough. What makes this a reference cycle? Also, why does this not cause the program to crash/freeze?

dlech replied:
A reference cycle leads to a memory leak. https://wiki.gnome.org/Projects/Vala/ReferenceHandling

So what happens here is that in the generated C code, it takes a reference to `volume_window` in the lambda expression and creates a closure structure. That closure is assigned to `volume_window` when the signal is connected, so now `volume_window` has a reference to itself and will never be freed.

So, you have to break this reference cycle. You could do this by disconnecting the signal, for example when the window is closed.

You can also do it by using a weak reference to `volume_window`. This way, when the C closure is created, it does not take a reference to `volume_window`. Since the signal will be disconnected when `volume_window` is destroyed, you don't have to worry about the signal being called after `volume_window` is freed.

WasabiFan replied:
Oh, I get it -- so it's not actively increasing the reference count in a cycle, it just holds a reference that will never let it be freed. I'll fix that.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+                volume_window.current_element.volume += VOLUME_STEP);
+
+            volume_window.volume_down.connect(() =>
+                volume_window.current_element.volume -= VOLUME_STEP);
```
reference cycle

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+                volume_window.current_element.volume -= VOLUME_STEP);
+
+            volume_window.volume_min.connect(() =>
+                volume_window.current_element.volume = IMixerElementViewModel.MIN_VOLUME);
```
reference cycle

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+            mixer_select_window.clear_elements();
+            // Re-initializing will return updated data, including volume
+            initialize_mixer();
+            for(MixerElement element = mixer.first_elem(); element != null; element = element.next()) {
```
space between `for` and `(`

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,110 @@
+        private bool _can_mute;
+        private bool _is_muted;
+
+        public string name {
```
why not `public string name { get; private set; }` instead?

Or if you want to be really glib-ish, `public string name { get; construct; }`

dlech replied:
nix the `construct`, I see now you have setters below. So, this should just be

```
public string name { get; set; }
```

There is no need for separate setter functions (unless they throw an error that you want to catch).

Also, there is no need to call `notify_property` in the setter. This is done automatically in GLib.

WasabiFan replied:
See above note; can't have a `set` clause because the interface defines it as get-only.

dlech replied:
Does `public override string name { get; set; }` work?

WasabiFan replied:
Whoops, forgot to reply to this one. Adding `override` does not resolve the problem; there isn't anything to override because the types don't match after adding a `set`.

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,110 @@
+            set {
+                _volume = int.min(100, int.max(0, value));
+
+                bool should_mute = _volume <= 0;
```
volume won't be < 0 here

WasabiFan replied:
This is one of those "intent" things; I do that to make it clear that the intent is to check if the volume is not at an audible level. It just makes the code easier to read at a glance, and more resilient if changes are made later.

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,110 @@
+            notify_property("is_muted");
+        }
+    }
+}
```
newline at end of file

WasabiFan replied:
👌

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+/*
+ * brickman -- Brick Manager for LEGO MINDSTORMS EV3/ev3dev
+ *
+ * Copyright 2014-2015 David Lechner <david@lechnology.com>
```
I didn't write this. :smile:

WasabiFan replied:
Whoops, thought I got them all; looks like I completely forgot to change this header.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+ * MA 02110-1301, USA.
+ */
+
+/* BatteryController.vala - Controller for monitoring battery */
```
This is not BatteryController

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,128 @@
+    public class MixerElementSelectorWindow : BrickManagerWindow {
+        Ui.Menu element_menu;
+
+        public signal void mixer_elem_selected (IMixerElementViewModel selected_element);
```
It wouldn't hurt to spell out `element` here for consistency since you haven't abbreviated it anywhere else.

WasabiFan replied:
Will change.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,128 @@
+
+        public signal void mixer_elem_selected (IMixerElementViewModel selected_element);
+
+        private class MixerElementWithSignalIds: Object {
```
This looks like it should be a `struct` rather than a `class`.

dlech replied:
There is probably a better way to handle signal disconnection, in which case this would not be needed anyway.

WasabiFan replied:
I spent a while trying to figure out how I wanted to do this, but in the end I decided that this was the least worst. This can't be a struct though, because I need to be able to jam it into `represented_object`.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,128 @@
+
+            // Update the menu item whenever the represented element changes
+            represented_object.notify_signal_handler_id = element.notify.connect((sender, property) => {
+                menu_item.label.text = get_element_label_text(element);
```
if you use a weak reference to `element` here, then you should not have to worry about disconnecting the signal later. It will be automatically disconnected when the `element` object is destroyed.

WasabiFan replied:
Ah, that should make it better.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,128 @@
+
+        public void clear_elements () {
+            var iter = element_menu.menu_item_iter ();
+            while (iter.size > 0)
```
`{}` around body of single-line `while` statements too.

WasabiFan replied:
👌

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,109 @@
+    public class MixerElementVolumeWindow : BrickManagerWindow {
+        public signal void volume_up ();
+        public signal void volume_down ();
+        public signal void volume_min ();
```
can we call this `mute` instead of `volume_min`?

WasabiFan replied:
Sure, I'll change that.

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,109 @@
+        public signal void volume_down ();
+        public signal void volume_min ();
+
+        IMixerElementViewModel _current_element = null;
```
No need for leading `_` (unless it conflicts with a public property name)

dlech replied:
heh, I see the property now.

dlech replied:
no need to initialize to null though.

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,109 @@
+            set {
+                _show_element_details = value;
+                update_from_element();
+            }
```
you can actually add

```
default = true;
```

instead of initializing the backing field above.

WasabiFan replied:
Is there an advantage to doing that over setting it on the backing field?

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,109 @@
+
+            var volume_min_item = new Ui.MenuItem("Mute");
+            volume_min_item.button.pressed.connect(() => {
+                if(_current_element != null)
```
when would we ever have this window without a `current_element`?

WasabiFan replied:
Between when it is created and when it is given an element, it will be null. But regardless of whether it _should_ be null, I'd like to design these classes so they won't throw unexpected errors even in the face of user error.

dlech commented on src/controller/AudioController.vala:
```diff
@@ -0,0 +1,99 @@
+
+        protected void initialize_mixer() {
+            mixer = null;
+            Mixer.open(out mixer);
```
Should be checking error values here in case something goes wrong. At a minimum:

```
err = Mixer.open (out mixer);
if (err < 0) {
    critical ("Opening mixer failed: %s", strerror (err));
    return;
}
```

WasabiFan replied:
Will-do.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,128 @@
+
+        protected string get_element_label_text(IMixerElementViewModel element) {
+            string mute_string = (element.can_mute && element.is_muted) ? ", muted" : "";
+            return "[%u] %s (%ld%%%s)".printf(element.index, element.name, element.volume, mute_string);
```
I don't think we need to display the index.

WasabiFan replied:
The index isn't just an array index; it is part of the identity of the element. You can have multiple devices with the same index, and you can have multiple devices with the same name. So we need to show the index otherwise it won't always be possible to tell the difference between two devices.

dlech replied:
Do you have an example of this? From what I have seen, the index is always `0`. Also, the index is not show in any other mixers that I have seen.

For example, here is `alsamixer` from my desktop computer, which I assume would be similar to what you would get if you had a USB sound card.

```
┌──────────────────────────────── AlsaMixer v1.1.0 ────────────────────────────────┐
│ Card: HDA ATI SB                                         F1:  Help               │
│ Chip: Realtek ALC892                                     F2:  System information │
│ View: F3:[Playback] F4: Capture  F5: All                 F6:  Select sound card  │
│ Item: Master [dB gain: 0.00]                             Esc: Exit               │
│   ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐   │
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   │
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   │
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   →
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   │
│   │▒▒│     │  │     │▒▒│     │▒▒│     │  │     │  │     │▒▒│     │▒▒│     │▒▒│   │
│   ├──┤     ├──┤     └──┘     ├──┤     ├──┤     └──┘     ├──┤     ├──┤     ├──┤   │
│   │OO│     │MM│              │OO│     │MM│              │OO│     │OO│     │OO│   │
│   └──┘     └──┘              └──┘     └──┘              └──┘     └──┘     └──┘   │
│   100      0<>0   100<>100 100<>100   0<>0     0<>0   100<>100   100      100    │
│< Master >Headphon   PCM     Front   Front Mi Front Mi Surround  Center    LFE    │
└──────────────────────────────────────────────────────────────────────────────────┘
```

WasabiFan replied:
The index is often 0, but not always. For example, from `amixer scontrols` on my machine:

```
Simple mixer control 'Master',0
Simple mixer control 'Headphone',0
Simple mixer control 'PCM',0
Simple mixer control 'Front',0
Simple mixer control 'Front Mic',0
Simple mixer control 'Front Mic Boost',0
Simple mixer control 'Surround',0
Simple mixer control 'Center',0
Simple mixer control 'LFE',0
Simple mixer control 'Side',0
Simple mixer control 'Line',0
Simple mixer control 'Line Boost',0
Simple mixer control 'IEC958',0
Simple mixer control 'IEC958 Default PCM',0
Simple mixer control 'Capture',0
Simple mixer control 'Capture',1
Simple mixer control 'Auto-Mute Mode',0
Simple mixer control 'Input Source',0
Simple mixer control 'Input Source',1
Simple mixer control 'Rear Mic',0
Simple mixer control 'Rear Mic Boost',0
```

Note how there are some of them that have duplicated names but different indices ("Capture" is one of them). As far as I can tell, they just ignore the fact that there can be (and are) some dupes in alsamixer.

---

It's also important to point out that, from what I understand, the things listed horizontally are not the mixer elements that we are operating on. They are enumerated channel controls within the mixer elements "Playback" or "Capture", which they have hard-coded. This seems fairly odd to me, as I know there are systems that have elements other than "Playback" and "Capture", but I guess they're commonplace enough that they decided to take a shortcut.

dlech replied:
It looks like what `alsamixer` does is that if index > 1, then show the index.

```
┌────────────────────────────── AlsaMixer v1.1.0 ──────────────────────────────┐
│ Card: HDA ATI SB                                     F1:  Help               │
│ Chip: Realtek ALC892                                 F2:  System information │
│ View: F3: Playback  F4:[Capture] F5: All             F6:  Select sound card  │
│ Item: Capture 1 [dB gain: -16.00, -16.00]            Esc: Exit               │
│                                                                              │
│     ┌──┐     ┌──┐     ┌──┐     ┌──┐     ┌──┐                       ┌──┐      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │  │     │  │     │  │                       │  │      │
│     │  │     │  │     │▒▒│     │  │     │  │                       │  │      │
│     │  │     │  │     │▒▒│     │  │     │  │                       │  │      │
│     │  │     │  │     │▒▒│     │  │     │▒▒│                       │  │      │
│     │  │     │  │     │▒▒│     │  │     │▒▒│                       │  │      │
│     └──┘     └──┘    L└──┘R    └──┘     └──┘   Front Mi Front Mi   └──┘      │
│                     CAPTURE  -------                                         │
│     0<>0     0<>0    40<>40    0<>0    24<>24                      0<>0      │
│   Front Mi Line Boo Capture <Capture >Digital  Input So Input So Rear Mic    │
└──────────────────────────────────────────────────────────────────────────────┘
```

I like that better than having a bunch of meaningless `[0]` on everything.

dlech replied:
> It's also important to point out that, from what I understand, the things listed horizontally are not the mixer elements that we are operating on. They are enumerated channel controls within the mixer elements "Playback" or "Capture", which they have hard-coded. This seems fairly odd to me, as I know there are systems that have elements other than "Playback" and "Capture", but I guess they're commonplace enough that they decided to take a shortcut.

You might call these "well known names". See http://www.alsa-project.org/~tiwai/writing-an-alsa-driver/ch06s03.html

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,105 @@
+                    return 0;
+                
+                // Do calculations as floats so avoid odd-looking increments from truncation
+                return (int)Math.round((volume - min_volume) * 100 / (float)(max_volume - min_volume));
```
I think we can simplify this by using `alsa_element.set_playback_volume_range(0, 100)`, then we can just write the value directly instead of scaling it ourselves. Haven't actually tried this out though.

dlech replied:
If it works, then calling `alsa_element.set_playback_volume_range()` would be more appropriate in the constructor.

dlech replied:
I have tested this now and it works very nicely.

WasabiFan replied:
This originally seemed like a great approach, however it isn't so good when you consider the fact that other apps can also set the range. So, if we did it this way, the `get` wouldn't work after using `alsamixer` (or another app) or the first time that brickman was used. At that point, these changes would only simplify the set portion; so it would add a call to set the range but remove the math on the next line. That seems like a wash to me.

dlech replied:
I see. I'm actually working on another app that pokes the alsa mixer, so this is good to know.

dlech replied:
By the way, I don't think using float has any benefit here. You are still truncating with `Math.round()`. As long as you multiply before dividing, you will get the exact same result.

dlech replied:
Another thing I just discovered is that if you call `snd_mixer_selem_set_playback_volume_all` (`MixerElement.set_playback_volume_all` in Vala), to the same value as the previous call, it does not actually change the mixer value if another mixer program has changed the value. I don't know that it is a problem with the code here, but just an interesting observation.

WasabiFan replied:
> By the way, I don't think using float has any benefit here. You are still truncating with  Math.round() . As long as you multiply before dividing, you will get the exact same result.

Unfortunately, doing it all as integer math (and multiplying before dividing) doesn't give the same result. I added the floats because both the write and the read were rounding down slightly through truncation so that the even increments on the 100 scale turned into inconsistent increments after the value was set and read. For example, going volume up then volume down repeatedly would continuously decrease the volume because of the rounding.

> Another thing I just discovered...

Interesting. I don't think there is much we can do about that, but it's good to know in case that turns into a problem later.

WasabiFan commented:
By the way, the desktop test spews a bunch of errors that  appear to be caused by not finding image assets. Not sure if you're already aware.

```
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_horizontal_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_vertical_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: NetworkConnectionMenuItem.vala:64: Could not find 'usb12x12.png'
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_horizontal_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_vertical_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: NetworkConnectionMenuItem.vala:64: Could not find 'wifi12x12.png'
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_horizontal_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: ev3dev_kit_ui_widget_set_vertical_align: assertion 'self != NULL' failed
** (brickman:17247): CRITICAL **: NetworkConnectionMenuItem.vala:64: Could not find 'bluetooth12x12.png'
```

WasabiFan committed:
Clean up sound code based on review feedback

WasabiFan commented:
Just updated my PR branch; I think I addressed everything, but there are enough comments that I may have missed things.

dlech commented:
> the desktop test spews a bunch of errors

I think you need to set the environment variable `XDG_DATA_DIRS=/path/to/build/dir/data` where `/path/to/build/dir` is the actual path to where you are running make and `data` is an automatically generated directory in the build directory.

Edit: if that doesn't work. `make && sudo make install`

dlech commented:
> I think I addressed everything

There are still many inconsistencies in style. I did not call all of them out individually. e.g. spaces before `(`, `{}` around statement bodies, etc.

WasabiFan commented:
> There are still many inconsistencies in style...

I've only addressed the ones you called out so far. I plan to go back in later today or tomorrow and change the rest as well.

WasabiFan committed:
Format audio code to follow spacing style conventions

WasabiFan committed:
Only show sound mixer index when nonzero

WasabiFan commented:
PR updated with style fixes and changes for other feedback.

WasabiFan commented:
Is there anything else you'd like me to do here?

dlech commented:
Oh probably. I haven't taken this for a test drive yet. I've been away for the last few days. Should be able to have a look again this weekend.

WasabiFan commented:
OK, that's totally fine. I just wanted to make sure that there wasn't something you were waiting for me to do. No rush.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,109 @@
+ * MA 02110-1301, USA.
+ */
+
+/* AlsaBackedMixerElement.vala - Implementation of IMixerElementViewModel using ALSA API */
```
If we are going to have view models now, let's start a new subdirectory for them., namely `view_model, and put this file in it.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,109 @@
+    public class AlsaBackedMixerElement: IMixerElementViewModel, Object {
+        const SimpleChannelId primary_channel_id = SimpleChannelId.MONO;
+
+        unowned MixerElement alsa_element;
```
Why is this unowned? It seems like we would want to keep a reference here or take ownership if it is not reference counted.

WasabiFan replied:
I tried to transfer the reference in the constructor of this class, but it caused errors that I was unable to resolve. I can repro the errors later today.

dlech commented on src/AlsaBackedMixerElement.vala:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        protected void set_is_muted (bool is_muted) {
+            if (can_mute)
```
`{}` around if statement body

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,111 @@
+    public interface IMixerElementViewModel : Object {
+        public const int MIN_VOLUME = 0;
+        public const int MAX_VOLUME = 100;
+        public const int HALF_VOLUME = (MAX_VOLUME + MIN_VOLUME) / 2;
```
Shouldn't this be `-` instead of `+`?

WasabiFan replied:
I'm taking the mean, so it sums the values then devides by the number of values.

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,111 @@
+ * MA 02110-1301, USA.
+ */
+
+/* AlsaInterface.vala - Definitions for interfacing with ALSA */
```
Same with this file. It should go in the `view_model` subdirectory.

dlech commented on src/AlsaInterface.vala:
```diff
@@ -0,0 +1,111 @@
+        public abstract bool is_muted { get; }
+    }
+
+    public class FakeMixerElement: IMixerElementViewModel, Object {
```
I think this class should go in a separate file under the `test` subdirectory so that it is not compiled in the "regular" version of Brickman.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+        public signal void mixer_element_selected (IMixerElementViewModel selected_element);
+
+        public MixerElementSelectorWindow () {
+            title = "Sound mixer elements";
```
I think we should just make the title "Volume Controls" since that is the only mixer element we can handle.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        protected string get_element_label_text (IMixerElementViewModel element) {
+            string mute_string = (element.can_mute && element.is_muted) ? ", muted" : "";
```
I think it would be better to use a string builder here. e.g.

```
var builder = new StringBuilder ();
builder.append (element.name);
if (element.index != 0) {
    element.append_printf ("[%u]", element.index);
}
if (element.can_mute && element.is_muted) {
    builder.append (" (muted)");
} else {
    builder.append_printf (" (%ld%%)", element.volume);
}

return builder.str;
```

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+            // the item in the correct place instead of sorting the entire list
+            // each time an item is inserted.
+            element_menu.sort_menu_items ((a, b) => {
+                IMixerElementViewModel element_a = a.represented_object as IMixerElementViewModel;
```
can use `var` here and next line.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+                IMixerElementViewModel element_b = b.represented_object as IMixerElementViewModel;
+
+                // Group by name, and sort by index within the same name
+                if (element_a.name == element_b.name)
```
`{}` around if/else bodies

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+            weak IMixerElementViewModel weak_element = element;
+            // Update the menu item whenever the represented element changes
+            element.notify.connect ((sender, property) => {
+                menu_item.label.text = get_element_label_text (weak_element);
```
There is a reference cycle here. `menu_item` has a reference to `element` in `represented_object` and `element` has  a reference to `menu_item` in the signal closure.

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+
+        public IMixerElementViewModel? first_element {
+            get {
+                if (element_menu.menu_item_iter ().size <= 0)
```
`{}` around body

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+        Ui.Label element_label;
+
+        public MixerElementVolumeWindow () {
+            title = "Volume control";
```
Could we make the title `"%s Volume"` where %s is the element name, such as `Playback`.

WasabiFan replied:
The title boxes don't handle strings that are long enough to wrap to multiple lines very well, so I'd rather not risk that odd behavior.

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+
+            var volume_up_item = new Ui.MenuItem ("+ Volume up");
+            volume_up_item.button.pressed.connect (() => {
+                if (_current_element != null)
```
`{}` around if statement body

dlech replied:
Also better to use `current_element` here instead of `_current_element` so that `_current_element` is not reference by the closure.

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+
+            var volume_down_item = new Ui.MenuItem ("- Volume down");
+            volume_down_item.button.pressed.connect (() => {
+                if (_current_element != null)
```
`{}`

dlech replied:
again, `current_element` instead of `_current_element`

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+            var mute_item = new Ui.MenuItem ("Mute");
+            mute_item.button.pressed.connect (() => {
+                if (_current_element != null)
+                    mute ();
```
`{}`

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+            if (_current_element == null) {
+                element_label.text = "???";
+            }
+            else {
```
`} else {`

I know... I'm a :japanese_ogre:

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+                element_label.text = "???";
+            }
+            else {
+                string index_string = _current_element.index == 0 ? "" : " [%u]".printf(_current_element.index);
```
Another place for string builder.

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                    model.get_value (iter, ControlPanel.SoundMixerElementsColumn.USER_DATA, out user_data);
+
+                    var mixer_element = (FakeMixerElement)user_data.get_pointer ();
+                    if (mixer_element != null)
```
`{}`

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                mixer_select_window.show ());
+
+            (builder.get_object ("sound-volume-window-button") as Gtk.Button).clicked.connect (() => {
+                if (mixer_select_window.first_element == null)
```
`{}`

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                volume_window.current_element = mixer_select_window.first_element;
+                volume_window.show_element_details = false;
+                volume_window.show ();
+            }
```
`} else {`

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                
+                mixer_elems_liststore.set (iter, ControlPanel.SoundMixerElementsColumn.USER_DATA, mixer_element.ref ());
+            }
+            else {
```
`} else {`

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+         * already exist.
+         */
+        private void update_fake_element_from_liststore (Gtk.TreeIter iter, Gtk.ListStore mixer_elems_liststore) {
+            Value name = get_liststore_value (mixer_elems_liststore, iter, ControlPanel.SoundMixerElementsColumn.NAME);
```
can use `var` for all of the variables in this method

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+            if (iter == null) {
+                mixer_elems_liststore.foreach ((model, path, current_iter) => {
+                    Value user_data = get_liststore_value (mixer_elems_liststore, current_iter, ControlPanel.SoundMixerElementsColumn.USER_DATA);
+                    FakeMixerElement other_element = (FakeMixerElement)user_data.get_pointer ();
```
`var`

dlech commented:
On the mixer element view model, you have the properties `can_mute` and `is_muted`. However, in the view you only use these properties together as far as I can tell, so could be simplified into a single property. To explain it another way, I can check the "Can mute" checkbox in the desktop test control panel and it does not change the view, so that tells me this does not need to be part of the view model or else there is a problem with your view.

Also, the "Is muted" checkbox is disabled in the desktop test control panel. We should be able to change this unless `is_muted` is a write-only property of the view model (and I would like to avoid write-only properties if possible).

dlech commented on src/controller/SoundController.vala:
```diff
@@ -0,0 +1,120 @@
+            mixer_select_window = new MixerElementSelectorWindow ();
+
+            mixer_select_window.mixer_elem_selected.connect ((selected_element) => {
+                if (volume_window == null)
```
`{}`

dlech commented on src/view/MixerElementSelectorWindow.vala:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        public void remove_element (IMixerElementViewModel element) {
+            var menu_item = element_menu.find_menu_item<IMixerElementViewModel> (element, (menu_item, target_element) => {
```
Using `menu_item` as a parameter in the lambda expression shadows the `menu_item` local variable. Please rename one of them.

dlech commented on src/view/MixerElementVolumeWindow.vala:
```diff
@@ -0,0 +1,110 @@
+
+            var mute_item = new Ui.MenuItem ("Mute");
+            mute_item.button.pressed.connect (() => {
+                if (_current_element != null)
```
`current_element`

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+            });
+
+            // Propagate changes from liststore to views
+            mixer_elems_liststore.row_changed.connect ((path, iter) => {
```
reference cycle on `mixer_elems_liststore`, although it probably doesn't matter if we never free `mixer_elems_liststore`.

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+            });
+
+            // Wire up handlers for volume window signals
+            volume_window.volume_up.connect (() => {
```
reference cycle on `volume_window`.

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                update_liststore_for_element (mixer_elems_liststore, volume_window.current_element);
+            });
+
+            volume_window.volume_down.connect (() => {
```
reference cycle on `volume_window`.

dlech commented on test/controller/FakeSoundController.vala:
```diff
@@ -0,0 +1,262 @@
+                update_liststore_for_element (mixer_elems_liststore, volume_window.current_element);
+            });
+
+            volume_window.mute.connect (() => {
```
reference cycle on `volume_window`.

WasabiFan commented:
> On the mixer element view model, you have the properties can_mute and is_muted. However, in the view you only use these properties together as far as I can tell, so could be simplified into a single property. To explain it another way, I can check the "Can mute" checkbox in the desktop test control panel and it does not change the view, so that tells me this does not need to be part of the view model or else there is a problem with your view.

The `can_mute` property reflects the ability of the underlying mixer element to mute, so it does need to be accessible. The problem here is that the volume window is lacking the proper check of `can_mute`; if you try the same test on the selector window the "Can mute" toggle will affect the view. I'll fix that omission in the volume window.

> Also, the "Is muted" checkbox is disabled in the desktop test control panel. We should be able to change this unless is_muted is a write-only property of the view model (and I would like to avoid write-only properties if possible).

I disabled the `is_muted` checkbox when I removed the mute button because the mute value is dictated by the volume now, so they aren't independent. The only case where the mute toggle would be helpful is emulating situations where other programs have muted the volume while the volume itself is still nonzero; do you think I should re-enable it for this case?

WasabiFan commented:
I've commented on the few items that require discussion; I'll make these changes later today or tomorrow.

WasabiFan committed:
More clean-up of audio code

WasabiFan commented:
:new:

dlech commented:
I haven't been in brickman in quite a while, but I have thought about this a bit (I haven't looked at the code in a while, so I may be out of touch with reality but...). I'm still not a fan of using a view model in brickman. The reason being that view model makes sense for a framework like WPF where you have automatic data binding via the IDE. We don't have that here, so essentially, we are doing the data binding twice, once in the view model and again in the controller. I would rather just do it once.

WasabiFan commented:
I decided to create that view model class only after I had implemented a fair amount of functionality without it. Every time I would implement something new, whether it was a property in the test app or the volume controls in the main window, I found myself cringing at the code I was writing. Part of my unhappiness was definitely a result of the much uglier code required to get the desktop test properly configured, which was actually the first thing I implemented -- even before the real controls or display. But fundamentally, I decided that most of the things I didn't like were a result of the fact that I was trying to model controls which modified a set of individual devices without any class to represent a single device, which is something that goes against the fundamental idea of OOP. So, I locally committed what I had and modified the code to use the mixer class that's in there now. I can't remember the specifics, but after comparing the two versions I decided that the one which used the view model class was cleaner and shorter in most cases, so I continued with that version.

I have no idea if there are things I could've done better with the original to make it less ugly or if there were design properties I wasn't considering (after all, I had never used GTK+ before, and the interaction between that and the GObject model was something that seriously confused/upset me) but I made a judgement call between the two options nonetheless. To a large extent, I chose the direction that was the most common in the OOP projects that I have built and been involved in, because I have a deep understanding of its benefits and pitfalls at the "design pattern" level.

If you are unhappy with the way I implemented this, I have no issue with you taking whatever parts of it you think are useful and re-structuring it to remove the view model. However, at this point I've moved my focus to other ev3dev-related things (I'm currently working on drastically improving the ev3dev-lang-python README as well as planning a new structure for the ev3dev docs) so I don't have time to allocate to changing the structure of this code. If you don't decide to do it yourself, I _might_ be able to in a few weeks once I am happy with the state of these other things.

snaildos commented:
Yes, this is old, *comeone*? Can we get some updates? This looks amazing and should be ready for a push.
